File: 2018/4/b.php
<?php
$input = [];
while ($line = rtrim(fgets(STDIN))) {
    $input[] = $line;
}

sort($input);
$guards = [];
$active_id = null;
foreach ($input as $line) {
    preg_match('/^\[[0-9-]+-([0-9]+) ([0-9]+):([0-9]+)\] ([a-zG]+) (#([0-9]+))?/', $line, $matches);
    list($dummy, $day, $hour, $min, $type) = $matches;
    $active_id = $matches[6] ?? $active_id;
    
    if (!isset($guards[$active_id])) {
        $guards[$active_id] = [
            'time' => [$hour,$min],
            'sleep_time' => 0,
            'details' => []
        ];
    }
    
    $guard = &$guards[$active_id];
        
    if ('falls' === $type) {
        $guard['time'] = [$hour,$min];
        continue;
    }
    
    if ('wakes' === $type) {
        if ($hour < $guard['time'][0]) {
            $min += 64;
        }
        $guard['sleep_time']+= $min - $guard['time'][1];
        for ($i = $guard['time'][1] ; $i < $min ; $i++) {
            $m = $i % 60;
            if (!isset($guard['details'][$m])) {
                $guard['details'][$m] = 0;
            }
            $guard['details'][$m]++;
        }
    }
}

$selected_minute = null;
$max_times = 0;
$selected_id = null;
foreach ($guards as $guard_id => $guard_data) {
    foreach ($guard_data['details'] as $minute => $times) {
        if ($times > $max_times) {
            $max_times = $times;
            $selected_minute = $minute;
            $selected_id = $guard_id;
        }
    }
}

echo $selected_minute * $selected_id;

File: 2018/4/b.exs
defmodule A do
  def add_sleep(guard, amount) do
    Map.update(guard, :sleep, amount, &(&1 + amount))
  end

  def add_details(guard, details) do
    Map.update(guard, :details, details, fn d ->
      Map.merge(d, details, fn _k, a, b -> a + b end)
    end)
  end

  def parse(str) do
    Regex.named_captures(
      ~r/\[[0-9]+-[0-9]+-[0-9]+ [0-9]+:(?<min>[0-9]+)\] (?<action>[a-zA-Z]+) (#(?<id>[\d]+))?/,
      str
    )
  end

  def add_event(%{"action" => "Guard", "id" => id}, acc) do
    Map.put(acc, :id, String.to_integer(id))
  end

  def add_event(%{"action" => "falls", "min" => min}, acc) do
    Map.put(acc, :time, String.to_integer(min))
  end

  def add_event(%{"action" => "wakes", "min" => min}, acc) do
    min = String.to_integer(min)
    sleep_time = min - acc.time
    details = Map.new(acc.time..(min - 1), fn x -> {x, 1} end)

    guards =
      Map.update(acc.guards, acc.id, %{id: acc.id, sleep: sleep_time, details: details}, fn g ->
        add_sleep(g, sleep_time) |> add_details(details)
      end)

    Map.put(acc, :guards, guards)
  end

  def resolve(input) do
    input |> resolve(%{guards: %{}})
  end

  def resolve([head | tail], acc) do
    acc =
      head
      |> parse
      |> add_event(acc)

    resolve(tail, acc)
  end

  def resolve([], acc) do
    {id, {min, _}} =
      acc.guards
      |> Enum.map(fn {id, guard} ->
        {id, guard.details |> Enum.max_by(fn {_min, amount} -> amount end)}
      end)
      |> Enum.max_by(fn {_id, {_min, amount}} -> amount end)

    id * min
  end
end

File.read!("input")
|> String.split(["\r", "\n"], trim: true)
|> Enum.sort()
|> A.resolve()
|> IO.inspect()
